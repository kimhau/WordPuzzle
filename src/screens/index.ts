export { default as Congratz } from './Congratz/Congratz';
export { default as Game } from './Game/Game';
export { default as Home } from './Home/Home';
export { default as Leaderboard } from './Leaderboard/Leaderboard';
export { default as Startup } from './Startup/Startup';
