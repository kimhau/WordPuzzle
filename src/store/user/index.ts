import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'user',
  initialState: { name: null, points: 0 } as UserState,
  reducers: {
    updateUser: (state, { payload: { name, points } }: UserPayload) => {
      if (typeof name !== 'undefined') {
        state.name = name;
      }
      if (typeof points !== 'undefined') {
        state.points = points;
      }
    },
    updatePoints: (state, { payload: { points } }: UserPayload) => {
      if (typeof points !== 'undefined') {
        state.points += points;
      }
    },
  },
});

export const { updateUser, updatePoints } = slice.actions;

export default slice.reducer;

export type UserState = {
  name: string | null;
  points: number;
};

export type UserPayload = {
  payload: Partial<UserState>;
};
