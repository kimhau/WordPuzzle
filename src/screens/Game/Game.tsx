import { faker } from '@faker-js/faker';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { SafeAreaView, Text } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import Grid from 'WordPuzzle/src/components/Grid/Grid';
import { useTheme } from 'WordPuzzle/src/hooks';
import { GameProps } from '../../../@types/navigation';

export interface RandomChar {
  char: string;
  used: boolean;
}

const Game = ({ navigation, route }: GameProps) => {
  const { category } = route.params;
  const [answer, setAnswer] = useState<string>('');
  const [attempAnswer, setAttempAnswer] = useState<string[]>();
  const [randomCharArr, setRandomCharArr] = useState<RandomChar[]>();

  const initialRandomCharArr = useMemo(
    () => answer.split('').sort(() => Math.random() - 0.5),
    [answer],
  );

  const createAnswer = useCallback(() => {
    let ans = '';
    switch (category) {
      case 'animals':
        ans = faker.animal.type().toUpperCase();
        break;
      case 'countries':
        ans = faker.address.country().toUpperCase();
        break;
      case 'chemicals':
        ans = faker.science.chemicalElement().name.toUpperCase();
        break;
      default:
        ans = faker.animal.type().toUpperCase();
        break;
    }
    setAnswer(ans);
  }, []);

  useEffect(() => {
    setAttempAnswer(answer.split('').map(() => '_'));
  }, [answer]);

  useEffect(() => {
    createAnswer();
  }, []);

  useEffect(() => {
    setRandomCharArr(
      initialRandomCharArr.map(char => {
        return { char, used: false };
      }),
    );
  }, [initialRandomCharArr]);

  useEffect(() => {
    if (attempAnswer && attempAnswer.join('') === answer.toUpperCase()) {
      const mutiplier =
        category === 'animals' ? 1 : category === 'countries' ? 2 : 3;
      navigation.reset({
        index: 1,
        routes: [
          {
            name: 'Home',
          },
          {
            name: 'Congratz',
            params: { points: answer.length * mutiplier, category },
          },
        ],
      });
    }
  }, [attempAnswer, answer, category]);

  const { Layout, Fonts, Common, Gutters } = useTheme();

  function onRandomCharPress(letter: string, index: number) {
    // const attempAnswerIndex = attempAnswer.indexOf('_');
    // if (attempAnswerIndex > -1) {
    setRandomCharArr(prev => {
      return prev.map((char, i) => {
        if (i === index) {
          return { ...char, used: true };
        }
        return char;
      });
    });
    setAttempAnswer(prev => {
      const attempAnswerIndex = prev.indexOf('_');
      if (attempAnswerIndex === -1) {
        return prev;
      }
      return prev.map((char, i) => {
        if (i === attempAnswerIndex) {
          return letter;
        }
        return char;
      });
    });
    // }
  }

  function onRemoveAttempAnswer(letter: string, index: number) {
    setRandomCharArr(prev => {
      const randomCharIndex = prev.findIndex(
        char => char.char === letter && char.used,
      );
      if (randomCharIndex === -1) {
        return prev;
      }
      return prev.map((char, i) => {
        if (i === randomCharIndex) {
          return { ...char, used: false };
        }
        return char;
      });
    });

    setAttempAnswer(prev => {
      return prev.map((char, i) => {
        if (i === index) {
          return '_';
        }
        return char;
      });
    });
  }

  return (
    <SafeAreaView style={[Layout.fill, Layout.scrollSpaceBetween]}>
      <TouchableOpacity onPress={() => navigation.goBack()}>
        <Text style={[Fonts.textRegular, Fonts.textPrimary]}>{'< Back'}</Text>
      </TouchableOpacity>
      <Grid data={attempAnswer} onPress={onRemoveAttempAnswer} />
      <Grid data={randomCharArr} onPress={onRandomCharPress} />
      <TouchableOpacity
        style={[
          Common.button.rounded,
          Gutters.largeBMargin,
          Gutters.regularHMargin,
        ]}
        onPress={createAnswer}
      >
        <Text style={Fonts.textCenter}>Skip</Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default Game;
