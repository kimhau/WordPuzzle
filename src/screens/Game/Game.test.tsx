import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from 'WordPuzzle/src/store';
import Game from './Game';

jest.mock('@faker-js/faker', () => ({
  faker: {
    animal: {
      type: jest.fn(() => 'lion'),
    },
  },
}));

//mock Math.random()
const randomFn = jest.spyOn(global.Math, 'random').mockReturnValue(0.5);

describe('Game', () => {
  const mockNavigation = {
    goBack: jest.fn(),
    reset: jest.fn(),
  };
  const mockRoute = {
    params: { category: 'animals' },
  };

  let component;

  beforeEach(() => {
    component = (
      <Provider store={store}>
        <Game navigation={mockNavigation} route={mockRoute} />
      </Provider>
    );
  });

  it('renders correctly', () => {
    const { toJSON } = render(component);
    expect(toJSON()).toMatchSnapshot();
  });

  it('allows user to enter letters and remove them', () => {
    const { getByTestId, getByText, getAllByText } = render(component);

    // Click on a random character and verify that it appears in the answer
    const randomCharL = getByText('L');
    const randomCharI = getByText('I');
    const randomCharO = getByText('O');
    const randomCharN = getByText('N');

    const answerChars = getAllByText('_');
    fireEvent.press(randomCharL);
    expect(answerChars[0].props.children).toEqual('L');

    // Click on the answer character and verify that it disappears
    fireEvent.press(answerChars[0]);
    expect(answerChars[0].props.children).toEqual('_');
  });

  it('allows user to skip the game', () => {
    const { getByText } = render(component);

    // Click on the Skip button and verify that navigation.reset is called
    const skipButton = getByText('Skip');
    fireEvent.press(skipButton);
    expect(randomFn).toHaveBeenCalled();
  });
});
