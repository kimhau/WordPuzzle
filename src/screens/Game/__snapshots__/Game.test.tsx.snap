// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
      {
        "flexGrow": 1,
        "justifyContent": "space-between",
      },
    ]
  }
>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#212529",
              "fontSize": 20,
            },
            {
              "color": "#E14032",
            },
          ]
        }
      >
        &lt; Back
      </Text>
    </View>
  </RNGestureHandlerButton>
  <RCTScrollView
    contentContainerStyle={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
    data={
      [
        "_",
        "_",
        "_",
        "_",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                _
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                _
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                _
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                _
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    contentContainerStyle={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
    data={
      [
        {
          "char": "L",
          "used": false,
        },
        {
          "char": "I",
          "used": false,
        },
        {
          "char": "O",
          "used": false,
        },
        {
          "char": "N",
          "used": false,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                L
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                I
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                O
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 149,
                "opacity": 1,
                "padding": 10,
                "width": 149,
              }
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 10,
                    "color": "#212529",
                    "height": 45,
                  },
                  {
                    "flex": 1,
                  },
                  false,
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "flex": 1,
                    },
                    {
                      "color": "#E14032",
                    },
                    {
                      "color": "#212529",
                      "fontSize": 40,
                    },
                  ]
                }
              >
                N
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={2}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E14032",
          "borderRadius": 10,
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 60,
          "marginHorizontal": 30,
          "opacity": 1,
          "paddingHorizontal": 30,
        }
      }
    >
      <Text
        style={
          {
            "textAlign": "center",
          }
        }
      >
        Skip
      </Text>
    </View>
  </RNGestureHandlerButton>
</RCTSafeAreaView>
`;
